ecp-provider/provider.c

```diff
@@ -0,0 +1,708 @@
+// Copyright 2023 Google LLC this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+#include <openssl/core.h>
+#include <openssl/core_dispatch.h>
+#include <openssl/core_names.h>
+#include <openssl/core_object.h>
+#include <openssl/crypto.h>
+#include <openssl/evp.h>
+#include <openssl/params.h>
+#include <openssl/provider.h>
+#include <openssl/ssl.h>
+#include <openssl/store.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <string.h>
+
+#ifdef _WIN32
+#define ECP_API __declspec(dllexport)
+#else
+#define ECP_API
+#endif
+
+#define debug_printf(X)                                                  \
+  do {                                                                   \
+    if (getenv("ENABLE_ENTERPRISE_CERTIFICATE_LOGS") != NULL) printf(X); \
+  } while (0)
+
+/* RSA signatures are much larger than EC. For EC 72 bytes is probably okay, but
+ * it depends on the EC curve. We will hard code to support a 4096 bit RSA key.
+ * This is currently the largest practical key for this use case. If the keys or
+ * signature get bigger, then this number should be increased to match the max
+ * possible signature ECP can create. This is calculated by taking a 4096 bit
+ * key, and converting it to bytes e.g. 512 bytes = 4096 bits / 8 bits per byte
+ */
+#define MAX_SIGNATURE_BUFFER_LEN 512
+
+typedef enum {
+  UNKNOWN_KEY_TYPE,
+  EC_KEY_TYPE,
+  RSA_KEY_TYPE,
+} key_types_t;
+
+const char provider_version[] = {"0.1.0"};
+const char provider_name[] = {"ECP Provider"};
+
+extern int SignForPython(char *configPath, const unsigned char *dig, int diglen,
+                         unsigned char *sig, size_t sigsize);
+extern int GetCertPemForPython(char *configPath, char *certHolder,
+                               int certHolderLen);
+extern char *GetKeyType(char *configPath);
+
+typedef struct ecp_context {
+  char *ecp_config_path;
+  size_t ecp_config_path_len;
+  char *digest_name;
+  size_t digest_name_len;
+  key_types_t key_type;
+  int initialized;
+  OSSL_LIB_CTX *libctx;
+} ecp_context_t;
+
+static int config_helper(ecp_context_t *ctx, char **ret_config) {
+  char *config = getenv("GOOGLE_API_CERTIFICATE_CONFIG");
+  if (config == NULL) {
+    if (ctx->ecp_config_path == NULL) {
```

Comment:
In this case I want to avoid hard coding paths, since it is a PITA in C

Reply:
Ok, we don't have to hard code a well known path, but can we at least swap the ordering of checking for ecp_config_path and the env var to be consistent? See https://github.com/googleapis/enterprise-certificate-proxy/blob/main/client/client.go#L152 (I know you had some previous reservations about the particular ordering, but the overarching principle in GUAC is to give precedence to explicitly passed in params to provide better isolation during runtime and also least surprise to the user in case the env var is set without their knowledge by a separate program.)